rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2168 @@
+        let foo = unsafe_transmute::<&mut [u8; 1024], &mut Foo, NeglectAlignment>(&mut array);
+    };
+
+    foo.a += 1;
+}
+```
+
+#### No Unsafe Needed!
+Better yet, we may use a completely *safe* transmute if we statically align the byte array:
+
+```rust
+use core::transmute::{
+    TransmuteInto,
+    options::NeglectAlignment,
+    stability::FixedLayout,
+};
+
+
+#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(C)]
+struct Foo {
+    a: i32,
+    b: i32,
+}
+
+#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(C, align(4)]
+struct Aligned([u8; 1024]);
+
+fn main() {
+    let mut array: Aligned = Aligned([1; 1024]);
+
+    let foo : &mut Foo = (&mut array[0]).transmute_into();
+
+    foo.a += 1;
+}
+```
+
+### Case Study: Abstractions for Packet Parsing
+***TODO: josh***
+
+### Case Study: Abstractions for Pointer Bitpacking
+[case-study-alignment]: #case-study-abstractions-for-pointer-bitpacking
+
+A [previous motivating example](#expressing-layout-requirements) involved a generic abstraction only usable with types meeting certain layout requirements: pointer bitpacking. Using this RFC's mechanisms, we can require that a reference to a generic type `T` has alignment of at least a given value (e.g., `8`) by first defining a ZST with that alignment:
+```rust
+#[derive(PromiseTransmutableFrom)]
+#[repr(align(8)]
+struct Aligned8;
+```
+and then using this `where` bound:
+```rust
+where
+    &T: TransmuteInto<&Aligned8>
+```
+
+## Rationale: `TransmuteFrom`/`TransmuteInto`
+
+### Why support arbitrary transmutation?
+Some [prior art][prior-art], especially in the crate ecosystem, provides an API that [only supports transmutations involving particular types](#Source-and-Destination-Types-Supported) (e.g., from/into bytes). As we discuss in the [prior art][prior-art] section, we believe that the inflexibility of such approaches make them a poor basis of a language proposal. In particular, these restrictive approaches don't leave room to grow: supporting additional transmutations requires additional traits.
+
+The API advocated by this proposal is unopinionated about what transmutations users might wish to do, and what transmutations the compiler is able to reason about. The implementation of this RFC may be initially very simple (and perhaps support no more than the restrictive approaches allow for), but then subsequently grow in sophistication—*without* necessitating public API changes.
+
+### Why *two* traits?
+If `TransmuteInto` is implemented in terms of `TransmuteFrom`, why provide it at all? We do so for consistency with libcore's [`From`/`Into`](https://doc.rust-lang.org/stable/rust-by-example/conversion/from_into.html) traits, and because directionality conveys intent: `TransmuteInto` connotes *conversion*, whereas `TransmuteInto` connotes initialization. We believe that the supporting code examples of this RFC demonstrate the explanatory benefits of providing *both* traits.
```

Comment:
I believe this should read:

`TransmuteInto` connotes *conversion*, whereas `TransmuteFrom` connotes initialization

With the second one being `TransmuteFrom`.

Reply:
```suggestion
If `TransmuteInto` is implemented in terms of `TransmuteFrom`, why provide it at all? We do so for consistency with libcore's [`From`/`Into`](https://doc.rust-lang.org/stable/rust-by-example/conversion/from_into.html) traits, and because directionality conveys intent: `TransmuteFrom` connotes *conversion*, whereas `TransmuteInto` connotes initialization. We believe that the supporting code examples of this RFC demonstrate the explanatory benefits of providing *both* traits.
```